### ISSUE TITLE
hard-coded paths in cobbler prevent installing in /usr/local or other directories

### ISSUE THREAD

**Issue by aaron-baidu** [cobbler/cobbler#2122]
There seems to be a lot of hard-coded paths in both the Makefile and the setup.py script for installing cobbler into the bare system (/var/lib/cobbler, /usr/bin/cobbler, etc). Setting an environment variable of DESTDIR to '/usr/local' allows the setup.py script to install to '/usr/local/usr/bin/' and '/usr/local/var/lib' instead of '/usr/local/bin' and so forth, merely appending a root-based path on top of any DESTDIR variable (which won't be in my path by default). In addition, the 'clean' script and parts of the 'install' script ignore the DESTDIR and try to update files in hardcoded paths.

I'd be happy to upload a fix for these if the problem is merely the Makefile and setup.py file, but I suspect the hard-coded paths are going to be a problem even after install?

**Comment by SchoolGuy** [comment]
@aaron-baidu We are aware of this problem. If you can help in any way we would appreciate this!

**Comment by aaron-baidu** [comment]
Will do. I'll make a PR, but from my personal account.

**Comment by mercury00** [comment]
Since this will break hard-coded paths in cobbler python after install, I'll have to take one step back and update the python to accept customized vars from settings. Looks like a number of fixme tags in the code already. Will start with a few modest targets.

**Comment by SchoolGuy** [comment]
This sounds very great!

**Comment by kwizart** [comment]
For information DESTDIR variable is used for packager maintainer to install into a separate directory (as a user). If you install by hand, you should modify the PREFIX variable to be /usr/local instead of /usr.
But then you need to modify others variables as appropriate.

**Comment by mercury00** [comment]
Finally had some time to dive into this and have been working on fixes.

My first pass is that I've identified all of the hard-coded paths in the cobbler python code.

These are being replaced with reference/api calls to vars in cobbler settings. In important places this is somewhat trivial - api, cobblerd, and cli are already able to access the cobbler settings dictionary. Other cobbler python code will need some careful care to do correctly.
For now, this will be done to merely preserve all the current functionality (no changing of paths or anything). I've needed to add the appropriate vars to cobbler settings where missing - namely, data_dir (as /var/lib/cobbler) and sysconf_dir (as /etc/cobbler). Some details about that last one I'll post about later.
However this will not yet address code outside of cobbler python - one of those is the django/web code and the other is the install/makefile code. That will have to be done after verifying the current edits do what they're supposed to - the second pass of edits. Just a heads up, HTH

**Comment by mercury00** [comment]
> For information DESTDIR variable is used for packager maintainer to install into a separate directory (as a user). If you install by hand, you should modify the PREFIX variable to be /usr/local instead of /usr.
> But then you need to modify others variables as appropriate.

Yes, indeed. I would need to do a little bit of editing of the setup.py script to allow some other parameters to be passed in (optional ones, with default settings) from the Makefile. 
I've edited the Makefile temporarily to at least substitute hard-coded /var/lib/cobbler instructions with a DATADIR var. Anyway, more on that later.

**Comment by SchoolGuy** [comment]
@mercury00 Thanks for the heads up! I would very much like if we could try to keep the cross-distro specfile up-to-date to have in the future a cross-package-build on a PR. I am already in discussion at work (The OBS guys are some nice friendly folks!) on how to do this.

**Comment by mercury00** [comment]
Yes, this sounds good to me.

It looks like much of the code in cobbler to determine the OS could actually be handled in the build process. That is, things having to do with OSes using different paths could rely on conf files (ie the settings file), which could be set at build time rather than polled at run time.

Anyway, I'm testing the sanity of my code changes piece by piece rather than in one fell swoop, (tried that, failed spectacularly), so, won't have a pull request until the first "stage" of changes is done.

--
Aaron
